backend/matching-service/src/utils/mq_utils.ts

```diff
@@ -0,0 +1,52 @@
+import { createMatch, MatchRequestItem } from "../handlers/matchHandler";
+import { isActiveRequest, isUserConnected } from "../handlers/websocketHandler";
+
+const matchingRequests = new Map<string, MatchRequestItem>();
+
+export const matchUsers = (newRequest: string) => {
+  const newRequestJson = JSON.parse(newRequest) as MatchRequestItem;
+  const newRequestUid = newRequestJson.user.id;
+  for (const [uid, pendingRequest] of matchingRequests) {
+    if (
+      isExpired(pendingRequest) ||
+      !isUserConnected(uid) ||
+      !isActiveRequest(uid, pendingRequest.id) ||
+      uid === newRequestUid
+    ) {
+      matchingRequests.delete(uid);
+      continue;
+    }
+    if (
+      isExpired(newRequestJson) ||
+      !isUserConnected(newRequestUid) ||
+      !isActiveRequest(newRequestUid, newRequestJson.id)
+    ) {
+      return;
+    }
+
+    if (isMatch(newRequestJson, pendingRequest)) {
+      matchingRequests.delete(uid);
+      createMatch(pendingRequest, newRequestJson);
+      return;
+    }
+  }
+  matchingRequests.set(newRequestUid, newRequestJson);
+};
+
+const isExpired = (data: MatchRequestItem): boolean => {
+  return Date.now() - data.sentTimestamp >= data.ttlInSecs * 1000;
+};
+
+const isMatch = (req1: MatchRequestItem, req2: MatchRequestItem): boolean => {
+  const hasCommonComplexity = req1.complexities.some((elem) =>
+    req2.complexities.includes(elem)
+  );
+  const hasCommonCategory = req1.categories.some((elem) =>
+    req2.categories.includes(elem)
+  );
+  const hasCommonLanguage = req1.languages.some((elem) =>
+    req2.languages.includes(elem)
+  );
+
+  return hasCommonComplexity && hasCommonCategory && hasCommonLanguage;
```

Comment:
was wondering if we shld be adding common lang now when thats a nice to have feature

Reply:
Oh I didn't realize it was a nice-to-have-feature, let me remove it thanks for spotting it~

Reply:
![image](https://github.com/user-attachments/assets/9ee846bd-1b69-4198-9c92-80f49686165d)

hmm i think in our specifications we mentioned we'll be matching based on all 3 criterias?

Reply:
ohhhh then i think can add back sorryy 😅 . i thot it was part of the nice to have where they can run the code...